helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -290,12 +296,20 @@ private void updateIdealRuleMap() {
   }
 
   public synchronized void refresh(HelixDataAccessor accessor) {
+    refresh(accessor, new HashSet<>());
+  }
+
+  /**
+   * @param accessor
+   * @param refreshedType Record the types that has been updated during the refresh.
+   */
+  protected synchronized void refresh(HelixDataAccessor accessor, Set<HelixConstants.ChangeType> refreshedType) {
     // Refresh raw data
     _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
-    refreshIdealState(accessor);
-    refreshLiveInstances(accessor);
-    refreshInstanceConfigs(accessor);
-    refreshResourceConfig(accessor);
+    refreshIdealState(accessor, refreshedType);
```

Comment:
I have some concerns about using an extra Set object to remember what has been refreshed. It's not about correctness or performance, but just feel it's redundant or unnecessary. I noticed the `refreshXXX` method is 99% same only differs in the type, so I guess we can avoid the extra set if we could do the refactor in the second phase with something like this :

```
Set<HelixConstants.ChangeType> refreshProperties(List<HelixConstants.ChangeType> properties) {
     Set<HelixConstants.ChangeType> refreshed = new HashSet<>();
     for (HelixConstants.ChangeType property : properties) {
          if (_propertyDataChangedMap.get(property).getAndSet(false)) {
                 _proepertyCacheMap.get(property).refresh(accessor);
                  LogUtil.logInfo(logger, getClusterEventId(), String
          .format("Reloaded %s for cluster %s, %s pipeline. Cnt: %s", property, _clusterName,
              getPipelineName(), _resourceConfigCache.getPropertyMap().keySet().size()));
          refreshed.add(property);
       }
     }
    return refreshed;
}
```
And the client who uses it like 
```
Set<HelixConstants.ChangeType> refreshed = refreshProperties(ImmutableList.of(xxxxx));
```

Reply:
I thought about this. I think the right way is making it property cache class method instead of in the data provider. But as you said, not in this PR.
The previous cache work is half done. We will need to finish the refactor work.